conda_forge_tick/migrators/libboost.py

```diff
@@ -12,44 +12,105 @@ def _slice_into_output_sections(meta_yaml_lines, attrs):
     sections = {}
     for i, line in enumerate(meta_yaml_lines):
-        # assumes order of names matches their appearance in meta_yaml,
-        # and that they appear literally (i.e. no {{...}}) and without quotes
-        if f"- name: {seek}" in line:
-            # found the beginning of the next output;
-            # everything until here belongs to the previous one
-            sections[prev] = meta_yaml_lines[pos:i]
-            # update
-            pos, prev, seek = i, seek, names.pop()
-            if seek == "dummy":
-                # reached the last output; it goes until the end of the file
-                sections[prev] = meta_yaml_lines[pos:]
-    if len(sections) != num_outputs + 1:
-        raise RuntimeError("Could not find all output sections in meta.yaml!")
-    return sections
-
-
-def _process_section(name, attrs, lines):
+        # we go through lines until we find the `outputs:` line
+        # we grab its position for later
+        if re_outputs_token.match(line) and outputs_token_pos is None:
+            # we cannot handle list-style outputs so raise an error
+            if re_outputs_token_list.match(line):
+                raise RuntimeError(
+                    "List-style outputs not supported for outputs slicing!"
+                )
+
+            outputs_token_pos = i
+            continue
+
+        # next we need to find the indent of the first output list element.
+        # it could be anything from the indent of the `outputs:` line to that number
+        # plus zero or more spaces. Typically, the indent is the same as the `outputs:`
+        # plus one of [0, 2, 4] spaces.
+        # ocne we find the first output list element, we build a regex to match any list
+        # element at that indent level.
+        if (
+            outputs_token_pos is not None
+            and re_output_start is None
+            and i > outputs_token_pos
+            and re_match_block_list.match(line)
+        ):
+            outputs_indent = len(line) - len(line.lstrip())
+            re_output_start = re.compile(r"^" + r" " * outputs_indent + r"-.*")
+
+        # finally we add slices for each output section as we find list elements at
+        # the correct indent level
+        # this block adds the previous output when it finds the start of the next one
+        if (
+            outputs_token_pos is not None
+            and re_output_start is not None
+            and re_output_start.match(line)
+        ):
+            sections[section_index] = meta_yaml_lines[pos:i]
+            section_index += 1
+            pos = i
+            continue
+
+    # the last output needs to be added
+    sections[section_index] = meta_yaml_lines[pos:]
+
+    # finally, if a block list at the same indent happens after the outputs section ends
+    # we'll have extra outputs that are not real. We remove them
+    # by checking if there is a name key in the dict
+    re_name = re.compile(r"^\s*(-\s+)?name:.*")
```

Comment:
I don't get the effort here. Why don't we just match for `- name: ...` directly, rather than determining the indent level, catching everything at that indent level, and _then_ matching on name?

Reply:
An output doesn't have to start with the name key as the first item in the dictionary. People do this as a custom but it is not required.

Reply:
> An output doesn't have to start with the name key as the first item in the dictionary.

Thanks, I hate it. 🙈 

I've _never_ seen that though (in contrast to templated outputs). In fact, I'd _want_ such feedstocks to break, because doing
```
outputs:
  - build:
      skip: true  # [not linux]
    name: ...
```
should just not be supported at all (i.e. we should turn it into a hard linter error) IMO.

I certainly don't think it's worth the implementation effort to accommodate such an extreme corner case.

Reply:
Yeah right now the libboost and cstdlib migrators are blocking perfectly good version and migration prs from being made because they don't handle templated names. 

We have one of these options

- turn off these minimigrators
- swallow their errors so they don't block PRs
- merge this pr to help fix things
- merge another pr to help fix things
- go around telling people the bot cannot handle templated output names

The implementation effort is fundamentally what is required to do the job at some standard, not what we are willing to give. The current recipe format allows a lot of edge cases and that generates complicated code in the bot. That's just how it is.

Reply:
There was a misunderstanding here. I agree we should fix templated names, no question about that (even though I'd like to get away from them wherever possible).

What I meant is that I don't think we should fix "name of output does not appear as first line", because the implementation here has IMO ballooned way further than what it is necessary to support templated outputs - which would mainly be to relax `f"- name: {seek}"` -> `f"- name:` (i.e. not include the package name), and incorporate your array/index logic of course.

If you feel the added generality for catching this extreme corner case[^1] is worth it, I won't stand in the way, but it makes me way less likely to touch that code again because I feel it's a bad trade-off and would have to resist the urge to remedy that.

[^1]: again, not talking about templated names here, but about not relying on having the hyphen before `name:`, i.e. `name:` being the first line per output

Reply:
That's fine. I'll take care of this code from now on. Thanks for the help in getting it started!

Reply:
We might also want to move this to be an "official" (albeit smith-internal) utility, rather than being tucked away in the boost migrator.